### DISCUSSION TITLE
Faikin ESP32S hardware

### DISCUSSION THREAD

**Discussion by dwmw2** [gekkekoe/esphome-ecodan-hp#31] [General]
Another suggestion for hardware that's fairly simple to use and doesn't involve much messing with connectors:
I've just set up one of the [ESP32-Faikin](https://github.com/revk/ESP32-Faikin) modules (bought from [Amazon UK](https://www.amazon.co.uk/dp/B0C2ZYXNYQ)). It comes preloaded with Daikin support, but runs esphome just fine.

![71HoJTxV2HL _AC_UL232_SR232,232_](https://github.com/user-attachments/assets/fcfba9c0-e648-44aa-a8e6-e3f4dbb92202)

You can get a premade lead which connects to the CN105 from [AliExpress](https://www.aliexpress.com/item/1005002904897793.html)

![f009945f-caaa-4f2a-9273-014ae081dbaa](https://github.com/user-attachments/assets/44e985df-02ad-4809-9add-7c36c6298c2d)

It's fairly simple to program it using probing pins in the USB or serial headers. My config (not yet attached to the heat pump) is at https://git.infradead.org/users/dwmw2/esp32-pool.git/blob/HEAD:faikin.yaml

**Comment by gekkekoe** [comment]
looks good, but a m5stack (~~5v tolerant~~) does the same (has nice casing, leds, reset buttons,..) and is only 8.50 euro instead of 30 pounds (=35 euro).
https://www.digikey.nl/en/products/detail/m5stack-technology-co-ltd/C124/18070571

Is there a compelling reason why you choose the faikin over the m5stack ?

**Reply by dwmw2** [reply]
No compelling reasons. I had bought a Faikin module before I even found esphome-ecodan-hp (I was looking at some of the other projects but hadn't found a ready-to-use esphome variant).

**Reply by gekkekoe** [reply]
I ordered one of these https://www.tindie.com/products/thehognl/heishamon-communication-pcb/
The connector is almost the same, but wires are a bit different. I will also port this project to run on this pcb, and probably use this for the proxy support testing (but still should work for other boards ofcourse). It has an optional utp which could be handy for ppl with bad wifi and 2 programmable relay switches)

**Reply by dwmw2** [reply]
Ooh, shiny! I'll order one of those too; thanks. I've already found another use for the AtomS3. 

We should probably make esphome-ecodan-hp use an underlying esphome [UART](https://esphome.io/components/uart.html) device instead of waggling pins directly. Then we should be able to use it on all kinds of boards and also in host mode too, shouldn't we? Like the [BL0942](https://esphome.io/components/sensor/bl0942.html) does, for example?

**Reply by gekkekoe** [reply]
I only support esp32 variants, those have at least 2 hardware uarts that we use directly, does not really make sense to use an abstraction layer on top of this. 

ESP 8266 won't be supported since I heed to rewrite the threads (and not really want to spent time on that, esp32 is plenty cheap).

**Reply by dwmw2** [reply]
Still learning esphome, but I think you're basically already using its UART driver instead of literally waggling pins directly, aren't you? It would just be a case of using an explicitly configured one instead of instantiating your own with pin numbers?

I'll take a look...

**Reply by gekkekoe** [reply]
https://github.com/gekkekoe/esphome-ecodan-hp/blob/ad98f9a9a8f5a9bf9d36628e50ef6a26e4530eea/components/ecodan/ecodan.h#L75
I use Serial1 which is one of the hardware uarts exposed via the esp lib. I think the  esphome uart sits on top of this
see https://esphome.io/api/uart__component__esp32__arduino_8h_source Line 53, it's just a wrapper on top of the HardwareSerial

**Reply by dwmw2** [reply]
This one also looks interesting, FWIW. https://www.tindie.com/products/jamesfowkes/quad-serial-breakout/ 

All four ports are 5v tolerant, so I could just use a single board (and esphome host mode) to connect to the three controllers plus the proxy for the MelCloud device.

**Reply by dwmw2** [reply]
Hm, the direct use of `Serial1` also seems to be the biggest thing which is stopping it from working with the esp-idf framework. I'm moving all my esphome boards to that at the moment because they get much better wifi connectivity.

**Reply by gekkekoe** [reply]
esp-idf was not working nicely with esphome last time I tried, not everything was working when I tried.

**Reply by dwmw2** [reply]
Yeah, I have some vague recollection of that too, which is why I had been using the arduino framework despite it being stuck in the 20th century and only supporting Legacy IP.

Since finding that *only* esp-idf could actually connect to the wifi when installing an unrelated [new device](https://git.infradead.org/?p=users/dwmw2/esp32-pool.git;a=commitdiff;h=87a704d0ab67797cd4e4072335c7c05d64a481da) last weekend though, I've converted everything over and it all seems to be working fine. I did have to [fix](https://github.com/TheStaticTurtle/esphome_syslog/pull/17) the syslog support but that's it.

**Reply by gekkekoe** [reply]
looks like with the recommended config
```esp32:
  board: esp32-s3-devkitc-1
  framework:
    type: esp-idf
    version: recommended
    # Custom sdkconfig options
    sdkconfig_options:
      COMPILER_OPTIMIZATION_SIZE: y
    # Advanced tweaking options
    advanced:
      ignore_efuse_mac_crc: false
```

it seems to build (esphome 2024.7.3) except for the hardware serial. in esp-idf thats done differently, hmm lets see how difficult it is to swap. I did all the ISR myself, would be happy if that's done in lib for me.

**Reply by dwmw2** [reply]
I am also happy to do some typing as well as just heckling; mostly just learning my way around at this point.  :)

**Reply by dwmw2** [reply]
I believe that for the AtomS3 Lite I did have to specify a newer version of the esp-idf framework (these specific versions randomly lifted from some GH issue about it):
```
esp32:
  board: esp32-s3-devkitc-1
  variant: esp32s3
  framework:
    type: esp-idf
    version: 5.0.2
    platform_version: 6.3.2
```

**Reply by gekkekoe** [reply]
happy to accept contributions, but this might be the thickest part

**Comment by gekkekoe** [comment]
https://www.hackster.io/news/raspberry-pi-unveils-the-pico-2-powered-by-the-dual-architecture-quad-core-rp2350-605253ab1235

async io, 5v tolerant. when the wifi version is available, it does look like a cool piece of hardware

**Comment by dwmw2** [comment]
Hm, I can't get the serial RX working. It works when connected to a USB serial port attached to my PC.
I run esphome-ecodan-hp in *host* mode from that same USB serial port attached to the heat pump, which works. 
But I'm just getting nonsense from the serial RX on the Faikin:
```
[16:00:52][I][ecodan.component:252]: Attempt to tx CONNECT_CMD!
[16:00:52][D][uart_debug:114]: >>> FC 5A 02 7A 03 CA 01 00 5C
[16:00:53][E][ecodan.component:046]: Dropping serial data; header magic mismatch
[16:00:53][D][uart_debug:114]: <<< 00 FF 00 FF 00 FF 00 FF 00 00 FF 01 00 00 FF 00 FF 00 FF 00 FF 00 00 00 FF 00 A5 4B CA 30 FF 00 FE 00 FF 00 00 FF 00 FF 00 01 FE 00 FE 00 00 FF 00 FE 00 00 FF 00 FF 00 FF 00 FF 00 FF 00 FF 00 FF 00 FF 00 FE 00 FF 00 FF 00 FF 00 FF 00 FF 00 00 FF 00 FF 00 00 00 FE 00 FF 00 FF 00 FE 00 FF 00 00 FF 01 FF 00 FF 00 FF 01 FF 00 FF 00 00 FF 00 FF 00 FF 00 FE 00 00
```

**Comment by gekkekoe** [comment]
try with some of my pin settings, heat pump acts differently

**Reply by dwmw2** [reply]
Yeah, tried copying those too. And experimenting with pullup/pulldown/invert. I'm back to the Arduino version now and trying to invert the pins to see if that works...

**Reply by gekkekoe** [reply]
try also recommended esp-idf, though both worked here

**Reply by dwmw2** [reply]
Ah, there's no pullup on the RX line. And I don't think doing it on the chip is sufficient; it's got to be on the outside of the protection circuit. I'll see if I've got a 10kΩ resistor lying around.

**Reply by dwmw2** [reply]
Yep, that works; I now have it working. Requested an onboard pullup in https://github.com/revk/ESP32-Faikin/issues/418
